### PR TITLE
portfolio-performance: new port

### PR DESCRIPTION
**Pull request by Janosch**
#### Description
New port for [Portfolio Performance](https://www.portfolio-performance.info/), a tool to calculate the performance of an investment portfolio.

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 11.1 20C69
Xcode 12.3.0.0.1.1607026830

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: finance/portfolio-performance/Portfile
Comment (harens) [inline finance/portfolio-performance/Portfile:16]:
```suggestion
depends_build       bin:mvn3:maven3 \
```

Only the binary is required as part of the build process, so I think it should be safe to specify it as a file dependency.

See https://guide.macports.org/chunked/reference.dependencies.html for more details.

File: finance/portfolio-performance/Portfile
Comment (harens) [inline finance/portfolio-performance/Portfile:17]: I'm presuming that since this is a Java project, Java would be required at both build time and runtime. I'm also going to assume that more than just the Java binary is required.

I would recommend using the Java PortGroup to set this all up for you. If you do so, this should be changed to a `depends_build-append`.

File: finance/portfolio-performance/Portfile
Comment (harens) [inline finance/portfolio-performance/Portfile:35]: Please correct me on this if I'm wrong on this, but the folder location makes it appear that this tool is built for x86_64. If this is the case, `supported_archs x86_64` should be added.

**Review by harens — CHANGES_REQUESTED** [review]
Thanks @Janosch for your contribution, as well as all your previous inputs to the MacPorts Project. They're greatly appreciated 👍.

Just a few small things from me. I've based my review on other maven ports such as [commons-text](https://github.com/macports/macports-ports/blob/749ad2787c30497020bdc4cca116eff9f09c800b/java/commons-text/Portfile) and [commons-lang3](https://github.com/macports/macports-ports/blob/749ad2787c30497020bdc4cca116eff9f09c800b/java/commons-lang3/Portfile), which you might find helpful to look through. Thanks again for submitting this port.

File: finance/portfolio-performance/Portfile
Comment (Janosch) [reply finance/portfolio-performance/Portfile:16]: AFAIK binary dependecies should be the exception, no? From the guide:

> But when satisfying a dependency with vendor-supplied software is preferred for special reasons, or when it is possible for more than one port to satisfy a dependency, then file dependencies may be used. 

Neither of those two mentioned cases apply in this situation, so I thought a port dependency makes more sense. No?

Comment (harens) [reply finance/portfolio-performance/Portfile:16]: Thanks @Janosch for your reply. I'd say that it's the second point:

> it is possible for more than one port to satisfy a dependency

There are many different maven ports that provide the required binary (e.g. [maven31](https://github.com/macports/macports-ports/blob/749ad2787c30497020bdc4cca116eff9f09c800b/java/maven31/Portfile), [maven32](https://github.com/macports/macports-ports/blob/749ad2787c30497020bdc4cca116eff9f09c800b/java/maven32/Portfile), and of course, [maven3](https://github.com/macports/macports-ports/blob/749ad2787c30497020bdc4cca116eff9f09c800b/java/maven3/Portfile)).

Comment (Janosch) [reply finance/portfolio-performance/Portfile:16]: Are you sure? Ony my machine it looks like this when I install maven32 and maven31 (and do the `port select --set maven`):
```
~/Dev/macports-ports ❯❯❯ ls -lah /opt/local/bin/mvn*
lrwxr-xr-x  1 root  wheel    20B 24 Jan 13:04 /opt/local/bin/mvn@ -> /opt/local/bin/mvn31
lrwxr-xr-x  1 root  admin    40B 24 Jan 13:04 /opt/local/bin/mvn31@ -> ../share/java/apache-maven-3.1.1/bin/mvn
lrwxr-xr-x  1 root  admin    29B 24 Jan 13:00 /opt/local/bin/mvn32@ -> ../share/java/maven32/bin/mvn
lrwxr-xr-x  1 root  wheel    25B 24 Jan 13:04 /opt/local/bin/mvnDebug@ -> /opt/local/bin/mvnDebug31
lrwxr-xr-x  1 root  admin    45B 24 Jan 13:04 /opt/local/bin/mvnDebug31@ -> ../share/java/apache-maven-3.1.1/bin/mvnDebug
lrwxr-xr-x  1 root  admin    34B 24 Jan 13:00 /opt/local/bin/mvnDebug32@ -> ../share/java/maven32/bin/mvnDebug
~/Dev/macports-ports ❯❯❯ mvn3
fish: Unknown command: mvn3
```
IMHO, only the port maven3 provides the binary `mvn3`. A different story would be the binary (or symlink)`mvn`. But in my case I need maven 3.x. so I cannot just use any mvn binary.

I can still change it if you insist but it feels like violating the macports guidelines.

File: finance/portfolio-performance/Portfile
Comment (Janosch) [reply finance/portfolio-performance/Portfile:17]: depends_build is definitly wrong for JDK, it should be depends_lib. I'll change that.

The java PG is currently broken for Big Sur. See https://trac.macports.org/ticket/61445. I would prefer not using it until this issue is resolved.

File: finance/portfolio-performance/Portfile
Comment (Janosch) [reply finance/portfolio-performance/Portfile:35]: Because of your comment I checked if it could be build for other archs. It doesent look like that other archs are supported from the upstream project. I will add your suggestion.

File: finance/portfolio-performance/Portfile
Comment (harens) [reply finance/portfolio-performance/Portfile:16]: I see. Thanks for testing that out, I didn't appreciate that although the `mvn` binary is provided, the `mvn3` binary isn't. Who would've known that? (You @Janosch, that's who 👍)

After doing some research, it seems that prior to Mavericks (10.9), maven3 came preinstalled. This might lead to the scenario "where the versions in macOS are often sufficient", but I don't know whether the `mvn3` binary is present or not.

**Comment by Janosch** [comment]
With the latest commit I implemented changes suggested from @harens I could agree with. One open point is whether to use the java port group or not.

Also, I added patches to remove the auto-update feature of portfolio-performance.

File: finance/portfolio-performance/Portfile
Comment (mf2k) [inline finance/portfolio-performance/Portfile:10]: You need to put braces around your GitHub handle and email address to link them. 

`maintainers         {@janosch mailbox.org:janosch1} openmaintainer`

**Review by mf2k — CHANGES_REQUESTED** [review]

File: finance/portfolio-performance/Portfile
Comment (Janosch) [reply finance/portfolio-performance/Portfile:10]: Good catch. Done.

File: finance/portfolio-performance/Portfile
Comment (Janosch) [reply finance/portfolio-performance/Portfile]: @harens Are you okay with not using the java PG for the reason state above? If not, I will change the port to use the java PG. 

I'd like to get this PR ready to merge.

Comment (harens) [reply finance/portfolio-performance/Portfile]: Apologies for the delay. Completely understand that you want to get this merged as soon as possible.

Other people might have different opinions on this, but assuming everything works at runtime, I think what you suggest is fine for the time being.

Having said that, once the issue with the Java PG is fixed, I'd prefer if we revert this. What happens if the user has openjdk11 installed? The [relevant README section](https://github.com/buchen/portfolio#building-with-maven) says this if fine, but the portfile requires openjdk15. What if they the have Oracle JDK? This is why there's a PortGroup to simplify this process.

If we stick with your suggestion for now, I'd prefer openjdk11 simply because it's an LTS release, unless there's any particular reason for going with openjdk15.

On a quick side-note, until we can use the Java PG, do you think it's worth setting `JAVA_HOME` in the scenario where the user has it pointing to a different SDK to the one in the portfile?

Thanks again for writing this portfile. A lot of great progress has been made so far.

_(Just reading through the trac ticket, I noticed you sent a PR with your suggestion for a workaround https://github.com/macports/macports-ports/pull/9617. Thank you for your work on this 👍.)_

Comment (Janosch) [reply finance/portfolio-performance/Portfile]: @harens I changed my mind regarding the java PG. As I have maven3 as dependency (which in turn uses the java PG), I will run into the java PG bug anyway so it does not really make sense to bypass it in the port of portfolio-performance. So with the latest push I added the java PG and changed `depends_build` to `depends_build-append` as requested.

**Comment by Janosch** [comment]
@harens @mf2k I resolved all changes requested so far and tested the app again. Please consider approval or let me know if you see some more issues.

**Review by mf2k — APPROVED** [review]

File: finance/portfolio-performance/Portfile
Comment (mf2k) [inline finance/portfolio-performance/Portfile:40]: The spacing is odd here. The braces don't need to follow the mode line. Also, the copy command should only be indented out 4 spaces.

**Review by mf2k — CHANGES_REQUESTED** [review]

File: finance/portfolio-performance/Portfile
Comment (Janosch) [reply finance/portfolio-performance/Portfile]: Thanks for the hint. It is fixed now.

File: finance/portfolio-performance/Portfile
Comment (harens) [inline finance/portfolio-performance/Portfile:24]:
```suggestion
java.version            11+
java.fallback           openjdk11
```

Can we allow later versions of Java? Reading throught the trac ticket, I see there are some issues with asterisk and plus notation, so please correct me if this wouldn't work with the current bug.

I would also suggest adding an LTS JDK port as a fallback.

Otherwise, lgtm. Thanks @Janosch.

**Review by harens — CHANGES_REQUESTED** [review]

File: finance/portfolio-performance/Portfile
Comment (Janosch) [reply finance/portfolio-performance/Portfile:24]: @harens It does not compile out of the box with higher Java versions. With Java 15, for example, I had to increase the version of _tycho_, a dependency of _portfolio-performance_. As the github page of _portfolio-performance_ states Java 11 explicitly as a prerequisite, I'd like to stick with 11 (without "+").

I added the java.fallback as suggested by you.

Comment (Janosch) [reply finance/portfolio-performance/Portfile:24]: And yes, 11+ does not work on Big Sur at the moment. It will install the fallback even if you have JDK 15 installed.

**Review by harens — APPROVED** [review]
Lgtm. Thanks for your work on this @Janosch, and for your suggestions on how to workaround the Java PG issue 👍.

**Review by mf2k — APPROVED** [review]

**Comment by harens** [comment]
Unless there's anything else, this PR has been waiting for a while and I think it looks good to merge.

**Comment by kencu** [comment]
multiple approvals. Looks clean. Passes all checks. Not much more we can do before committing.

**Pull request merged**
